### PR TITLE
Add outbound CallChannel support to Kotlin/Native

### PR DESCRIPTION
**Pull request by JakeWharton**
Refs #296.

Still many tests to port and exception handling to implement but those will be separate PRs.

### CODE REVIEW

File: zipline/build.gradle.kts [hunk 1]
```diff
@@ -158,6 +158,7 @@ cklib {
     srcDirs = project.files(file("native/quickjs"))
     compilerArgs.addAll(
       listOf(
+        //"-DDUMP_LEAKS=1", // For local testing ONLY!
```
Comment (swankjesse) [inline zipline/build.gradle.kts:161]: 👍🏻

File: zipline/src/engineTest/kotlin/app/cash/zipline/LoggingCallChannel.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.zipline
+
+import app.cash.zipline.internal.bridge.CallChannel
+
+class LoggingCallChannel : CallChannel {
```
Comment (swankjesse) [inline zipline/src/engineTest/kotlin/app/cash/zipline/LoggingCallChannel.kt:20]: at one point the thing like this I built caused recursive calls back into invoke() and that didn’t work

File: zipline/src/nativeInterop/cinterop/quickjs.def [hunk 3]
```diff
@@ -22,3 +22,41 @@ static inline int JsValueGetBool(JSValue v) {
...
+
+static inline JSValue JsTrue() {
+  return JS_TRUE;
+}
+
+static inline JSValue JsFalse() {
+  return JS_FALSE;
+}
+
+// Things that could be written in Kotlin but were easier in C
```
Comment (swankjesse) [inline zipline/src/nativeInterop/cinterop/quickjs.def:38]: Good comment

File: zipline/src/nativeMain/kotlin/app/cash/zipline/QuickJs.kt [hunk 4]
```diff
@@ -280,7 +299,79 @@ actual class QuickJs private constructor(
...
+      classDef.class_name = "OutboundCallChannel".cstr.ptr
+      JS_NewClass(runtime, id.value, classDef.ptr)
+
+      id.value.toInt() // Why doesn't JS_NewObjectClass accept a UInt / JSClassID?
+    }
+
+    val jsOutboundCallChannel = JS_NewObjectClass(context, outboundCallChannelClassId)
+    if (JS_IsException(jsOutboundCallChannel) != 0 ||
+        JS_SetProperty(context, globalThis, propertyName, jsOutboundCallChannel) <= 0) {
+      JS_FreeAtom(context, propertyName)
```
Comment (swankjesse) [inline zipline/src/nativeMain/kotlin/app/cash/zipline/QuickJs.kt:336]: put these in finally clauses? Couldn’t do that in the original C++

File: zipline/src/nativeMain/kotlin/app/cash/zipline/QuickJs.kt [hunk 5]
```diff
@@ -364,4 +455,77 @@ actual class QuickJs private constructor(
     }
     return array
   }
+
+  private fun Boolean.toJsValue(): CValue<JSValue> {
+    return if (this) JsTrue() else JsFalse()
```
Comment (swankjesse) [inline zipline/src/nativeMain/kotlin/app/cash/zipline/QuickJs.kt:458]: my first time seeing `if (this)` and it breaks my brain a bit

**Review by swankjesse — APPROVED** [review]

File: zipline/src/nativeMain/kotlin/app/cash/zipline/QuickJs.kt (hunk 4 above)
Comment (JakeWharton) [reply zipline/src/nativeMain/kotlin/app/cash/zipline/QuickJs.kt:336]: I actually had that once. I'm not sure why it was lost. This PR had _a lot_ of churn to get it working.

**Pull request merged**